### PR TITLE
Verify rpcap control socket status during reception

### PR DESCRIPTION
**Pull request by kevinboulain**
This is probably the first of a few patches so that a client can more reliably handle different connections failure.

In this patch, I chose to extend the `select` on the control socket during packet retrieval.
When using TCP, it's fine to totally avoid checking the state of the control socket (as if the data socket dies, the control socket would probably too), but when using UDP, the client may be stuck indefinitely even though the control socket surely died.

I'm not sure there is a better way than using `MSG_PEEK` (which should be safe in this case as the control socket seems to be discarded until the end of the packet capture). `sock_recv` didn't seem suitable (can't pass the flag, message oriented, etc).

The advantage of this solution being that TCP keep-alives  (probably via `SO_KEEPALIVE` and `TCP_USER_TIMEOUT`) could be set on the control socket to be more reactive when the connection drops (#766 is related but it isn't suitable for UDP and probably too hard-coded).

Do you have any idea on what would be the best way to provide keep-alives from the API standpoint?
It can think of two ways:
 - introduce a new flag, for example `PCAP_OPENFLAG_KEEPALIVE` but the downside is that we shouldn't probably hard-code timeouts 
 - introduce helpers functions, that could be used after the call to `pcap_open`

What do you think?

Regards.

**EDIT**: would have liked to use the `MAX` macro in the `select` but it's defined in `optimize.c` and there doesn't seem to have some internal/miscellaneous header.

### CODE REVIEW

File: pcap-rpcap.c [hunk 1]
```diff
@@ -413,8 +419,31 @@ static int pcap_read_nocb_remote(pcap_t *p, struct pcap_pkthdr *pkt_header, u_ch
...
 
+	/*
+	 * In the rpcap protocol, once the capture starts, the control socket isn't
+	 * used anymore until the capture ends.
+	 * However, it's the only way to check for connection errors
+	 * as the data socket may uses UDP.
+	 */
+	if (FD_ISSET(pr->rmt_sockctrl, &rfds)) {
+		uint8 byte;
+		const ssize_t nread = recv(pr->rmt_sockctrl, &byte, sizeof(byte), MSG_PEEK);
```
Comment (guyharris) [inline pcap-rpcap.c]: This should call `sock_recv()`, to filter out UN\*X vs. Windows differences.  Note that `sock_recv()` returns an `int`; `ssize_t` is a UN\*X type but not a Windows type.

Comment (ether42) [reply pcap-rpcap.c]: Yep, I just saw the builder failure and exported it in the header before seeing your comment.

The problem with `sock_recv` (as quickly explained above) is that I should add another parameter just to handle `MSG_PEEK` or a special flag and that there are already special flags conflicting with this use case (like `SOCK_EOF*`). That seemed less clean to me at first :)

Comment (guyharris) [reply pcap-rpcap.c]: >  I should add another parameter just to handle `MSG_PEEK` or a special flag

No, `SOCK_MSG_PEEK` already been added in d491c4d68d38902ce71fcceed49aeac1f574f533, so you don't have to add it, just use it.

File: pcap-rpcap.c [hunk 2]
```diff
@@ -413,8 +419,31 @@ static int pcap_read_nocb_remote(pcap_t *p, struct pcap_pkthdr *pkt_header, u_ch
...
+		if (nread <= 0)
+		{
+			if (nread == -1)
+				sock_geterror("recv(): ", p->errbuf, PCAP_ERRBUF_SIZE);
+			/* control socket has been properly closed or got an error, stop capture */
+			SOCK_DEBUG_MESSAGE("Control socket closed, stopping capture");
+			return -1;
```
[the remark this answers is not in the document]
Comment (guyharris) [reply pcap-rpcap.c]: `sock_recv()` already fills in the error message.

File: pcap-rpcap.c [hunk 3]
```diff
@@ -396,10 +396,16 @@ static int pcap_read_nocb_remote(pcap_t *p, struct pcap_pkthdr *pkt_header, u_ch
...
+	 *
+	 * While not strictly necessary, it's best to include the control socket.
+	 * It allows us to check for a connection drop as the data socket may use UDP
+	 * and as such, is without any mean to report back any error to the client.
 	 */
+	FD_SET(pr->rmt_sockctrl, &rfds);
 	FD_SET(pr->rmt_sockdata, &rfds);
 
-	retval = select((int) pr->rmt_sockdata + 1, &rfds, NULL, NULL, &tv);
+    const int max_fd = pr->rmt_sockctrl > pr->rmt_sockdata ? pr->rmt_sockctrl : pr->rmt_sockdata;
```
Comment (guyharris) [inline pcap-rpcap.c]: Indent this line the same way the adjacent line is indented.

File: pcap-rpcap.c (hunk 1 above)
Comment (ether42) [reply pcap-rpcap.c]: Thank you!

**Comment by ether42** [comment]
AppVeyor crashed two times on a network fetch, looks like it's building the code otherwise.

**Comment by infrastation** [comment]
Is there a good reason against rebasing and merging these changes?

**Pull request closed without merging**